clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -276,27 +329,83 @@ else if (expireMs <= now) {
             }
         }
 
-        String previousClusterId = cluster.clusterResource().clusterId();
+        String previousClusterId = cache.cluster().clusterResource().clusterId();
 
-        this.cluster = metadataResponse.cluster();
+        this.cache = handleMetadataResponse(metadataResponse, topic -> true);
         this.unavailableTopics = metadataResponse.unavailableTopics();
-
-        fireListeners(cluster, unavailableTopics);
+        Cluster clusterForListeners = this.cache.cluster();
+        fireListeners(clusterForListeners, unavailableTopics);
 
         if (this.needMetadataForAllTopics) {
             // the listener may change the interested topics, which could cause another metadata refresh.
             // If we have already fetched all topics, however, another fetch should be unnecessary.
             this.needUpdate = false;
-            this.cluster = metadataResponse.cluster(topics.keySet());
+            this.cache = handleMetadataResponse(metadataResponse, topics.keySet()::contains);
         }
 
-        String newClusterId = cluster.clusterResource().clusterId();
+        String newClusterId = cache.cluster().clusterResource().clusterId();
         if (newClusterId == null ? previousClusterId != null : !newClusterId.equals(previousClusterId))
             log.info("Cluster ID: {}", newClusterId);
-        clusterResourceListeners.onUpdate(cluster.clusterResource());
+        clusterResourceListeners.onUpdate(clusterForListeners.clusterResource());
 
         notifyAll();
-        log.debug("Updated cluster metadata version {} to {}", this.version, this.cluster);
+        log.debug("Updated cluster metadata version {} to {}", this.version, this.cache);
```

Comment:
Probably need to give MetadataCache a `toString`. Perhaps it can just delegate to `Cluster`.